sw/device/silicon_creator/lib/crypto/rsa_3072/rsa_3072.s

```diff
@@ -593,58 +595,29 @@ precomp_rr:
   /* w16 <= 1 */
   bn.addi   w16, w31, 1
 
-  /* Initialize c0
+  /* Initialize c0.
      c0 = [w4:w15] <= [w4:w16] >> 1 = 2^3701 */
   bn.rshi   w15, w16, w15 >> 1
 
-
-precomp_rr_sub_start:
-
-  /* Repeatedly subtract M until c0 < M.
-       [w16:w27] <= (c0 - M) */
-  jal       x1, subtract_modulus_var
-
-  /* Extract borrow bit from flags register. */
-  csrrs     x2, 0x7c0, x0
-  andi      x2, x2, 1
-
-  /* If borrow is set, then subtraction underflowed, meaning c0 < M; done. */
-  bne       x2, x0, precomp_rr_sub_done
-
-  /* If we got here, then c0 > M; set c0 = c0 - M and repeat.
-       c0 = [w4:w15] <= [w16:w27] = c0 - M */
-  li        x8, 4
-  li        x11, 16
-  loopi     12, 2
-    bn.movr   x8, x11++
-    addi      x8, x8, 1
-
-  /* Jump back to start of subtractions. */
-  beq       x0, x0, precomp_rr_sub_start
-
-precomp_rr_sub_done:
-
-  /* Now, we know that c0 = [w4:w15] = (2^3071) mod M. */
-
   /* One modular doubling to get c1 \equiv 2^3072 mod M.
      c1 = [w4:w15] <= ([w4:w15] * 2) mod M = (2^3072) mod M */
   jal     x1, double_mod_var
 
-  /* Compute (2^3072)^2 mod M by performing 3072 modular doublings.
-     Loop is nested only because #iterations must be < 1024 */
-  loopi     12, 4
-    loopi     256, 2
-      jal     x1, double_mod_var
-      /* Nop because inner loopi can't end on a jump instruction. */
-      nop
-    /* Nop because outer loopi can't end on a loop instruction. */
+  /* Set number of iterations for the loop.
+       x9 <= 3072 */
+  li      x9, 3072
```

Comment:
TBH, I would have folded this into the block below, unless there is a specific desire for `x9 <= 3072` to be visible? It looks like a scratch register... but it's your assembly, not mine. =)